File: LeonEdBackend/Services/Implementations/GradingService.cs
using Microsoft.EntityFrameworkCore;
using LeonEdBackend.Data;
using LeonEdBackend.DTOs.Common;
using LeonEdBackend.DTOs.GradingDtos;
using LeonEdBackend.Helpers;
using LeonEdBackend.Models;
using LeonEdBackend.Services.Interfaces;

namespace LeonEdBackend.Services.Implementations
{
    public class GradingService : IGradingService
    {
        private readonly AppDbContext _context;

        // Default grading rules if school hasn't configured custom ones
        private static readonly List<(Grade Grade, int Min, int Max, string Remark)> DefaultRules = new()
        {
            (Grade.A, 70, 100, "Excellent"),
            (Grade.B, 60, 69, "Very Good"),
            (Grade.C, 50, 59, "Good"),
            (Grade.D, 45, 49, "Fair"),
            (Grade.E, 40, 44, "Pass"),
            (Grade.F, 0, 39, "Fail")
        };

        public GradingService(AppDbContext context)
        {
            _context = context;
        }

        public async Task<ApiResponse<List<GradingRuleResponse>>> SetGradingRules(Guid schoolId, BulkCreateGradingRulesRequest request)
        {
            // Validate school exists
            var school = await _context.Schools.FindAsync(schoolId);
            if (school == null)
                return ApiResponse<List<GradingRuleResponse>>.FailResponse("School not found.");

            // Remove existing rules for this school
            var existingRules = await _context.GradingRules
                .Where(g => g.SchoolId == schoolId)
                .ToListAsync();
            _context.GradingRules.RemoveRange(existingRules);

            // Create new rules
            var newRules = request.Rules.Select(r => new GradingRule
            {
                Id = Guid.NewGuid(),
                SchoolId = schoolId,
                Grade = r.Grade,
                MinScore = r.MinScore,
                MaxScore = r.MaxScore,
                Remark = r.Remark
            }).ToList();

            _context.GradingRules.AddRange(newRules);
            await _context.SaveChangesAsync();

            var response = newRules.Select(r => new GradingRuleResponse
            {
                Id = r.Id,
                Grade = r.Grade.ToString(),
                MinScore = r.MinScore,
                MaxScore = r.MaxScore,
                Remark = r.Remark
            }).ToList();

            return ApiResponse<List<GradingRuleResponse>>.SuccessResponse(response, "Grading rules updated successfully.");
        }

        public async Task<ApiResponse<List<GradingRuleResponse>>> GetGradingRules(Guid schoolId)
        {
            var rules = await _context.GradingRules
                .Where(g => g.SchoolId == schoolId)
                .OrderByDescending(g => g.MaxScore)
                .ToListAsync();

            // Return defaults if no custom rules
            if (!rules.Any())
            {
                var defaults = DefaultRules.Select(r => new GradingRuleResponse
                {
                    Id = Guid.Empty,
                    Grade = r.Grade.ToString(),
                    MinScore = r.Min,
                    MaxScore = r.Max,
                    Remark = r.Remark
                }).ToList();

                return ApiResponse<List<GradingRuleResponse>>.SuccessResponse(defaults, "Default grading rules (no custom rules configured).");
            }

            var response = rules.Select(r => new GradingRuleResponse
            {
                Id = r.Id,
                Grade = r.Grade.ToString(),
                MinScore = r.MinScore,
                MaxScore = r.MaxScore,
                Remark = r.Remark
            }).ToList();

            return ApiResponse<List<GradingRuleResponse>>.SuccessResponse(response, "Grading rules retrieved.");
        }

        public async Task<Grade> GetGrade(Guid schoolId, decimal score)
        {
            var rules = await _context.GradingRules
                .Where(g => g.SchoolId == schoolId)
                .ToListAsync();

            int intScore = (int)Math.Round(score);

            if (rules.Any())
            {
                var match = rules.FirstOrDefault(r => intScore >= r.MinScore && intScore <= r.MaxScore);
                return match?.Grade ?? Grade.F;
            }

            // Use defaults
            var defaultMatch = DefaultRules.FirstOrDefault(r => intScore >= r.Min && intScore <= r.Max);
            return defaultMatch.Grade;
        }

        public async Task<string> GetRemark(Guid schoolId, decimal score)
        {
            var rules = await _context.GradingRules
                .Where(g => g.SchoolId == schoolId)
                .ToListAsync();

            int intScore = (int)Math.Round(score);

            if (rules.Any())
            {
                var match = rules.FirstOrDefault(r => intScore >= r.MinScore && intScore <= r.MaxScore);
                return match?.Remark ?? "Fail";
            }

            var defaultMatch = DefaultRules.FirstOrDefault(r => intScore >= r.Min && intScore <= r.Max);
            return defaultMatch.Remark;
        }
    }
}

File: src/services/grading.service.ts
import { prisma } from "../config/db";
import { successResponse, failResponse } from "../utils/response";

export class GradingService {
  private static readonly DefaultRules = [
    { grade: "A" as const, min: 70, max: 100, remark: "Excellent" },
    { grade: "B" as const, min: 60, max: 69, remark: "Very Good" },
    { grade: "C" as const, min: 50, max: 59, remark: "Good" },
    { grade: "D" as const, min: 45, max: 49, remark: "Fair" },
    { grade: "E" as const, min: 40, max: 44, remark: "Pass" },
    { grade: "F" as const, min: 0, max: 39, remark: "Fail" },
  ];

  static async setGradingRules(schoolId: string, request: any) {
    const school = await prisma.school.findUnique({
      where: { id: schoolId },
    });

    if (!school) {
      return failResponse("School not found.");
    }

    // Delete existing rules for this school, then insert new ones in transaction
    await prisma.$transaction([
      prisma.gradingRule.deleteMany({
        where: { schoolId },
      }),
      prisma.gradingRule.createMany({
        data: request.rules.map((r: any) => ({
          schoolId,
          grade: r.grade,
          minScore: r.minScore,
          maxScore: r.maxScore,
          remark: r.remark || "",
        })),
      }),
    ]);

    const newRules = await prisma.gradingRule.findMany({
      where: { schoolId },
    });

    const response = newRules.map((r) => ({
      id: r.id,
      grade: r.grade,
      minScore: r.minScore,
      maxScore: r.maxScore,
      remark: r.remark,
    }));

    return successResponse(response, "Grading rules updated successfully.");
  }

  static async getGradingRules(schoolId: string) {
    const rules = await prisma.gradingRule.findMany({
      where: { schoolId },
      orderBy: { maxScore: "desc" },
    });

    if (rules.length === 0) {
      const defaults = this.DefaultRules.map((r) => ({
        id: "00000000-0000-0000-0000-000000000000",
        grade: r.grade,
        minScore: r.min,
        maxScore: r.max,
        remark: r.remark,
      }));
      return successResponse(defaults, "Default grading rules (no custom rules configured).");
    }

    const response = rules.map((r) => ({
      id: r.id,
      grade: r.grade,
      minScore: r.minScore,
      maxScore: r.maxScore,
      remark: r.remark,
    }));

    return successResponse(response, "Grading rules retrieved.");
  }

  static async getGrade(schoolId: string, score: number): Promise<"A" | "B" | "C" | "D" | "E" | "F"> {
    const rules = await prisma.gradingRule.findMany({
      where: { schoolId },
    });

    const intScore = Math.round(score);

    if (rules.length > 0) {
      const match = rules.find((r) => intScore >= r.minScore && intScore <= r.maxScore);
      return (match?.grade as any) || "F";
    }

    const defaultMatch = this.DefaultRules.find((r) => intScore >= r.min && intScore <= r.max);
    return defaultMatch ? defaultMatch.grade : "F";
  }

  static async getRemark(schoolId: string, score: number): Promise<string> {
    const rules = await prisma.gradingRule.findMany({
      where: { schoolId },
    });

    const intScore = Math.round(score);

    if (rules.length > 0) {
      const match = rules.find((r) => intScore >= r.minScore && intScore <= r.maxScore);
      return match?.remark || "Fail";
    }

    const defaultMatch = this.DefaultRules.find((r) => intScore >= r.min && intScore <= r.max);
    return defaultMatch ? defaultMatch.remark : "Fail";
  }
}
